core/sequencer/beam_fns.go

```diff
@@ -0,0 +1,223 @@
+			emit(&ktpb.EntryUpdate{
+				Mutation:  m.Mutation,
+				Committed: m.ExtraData,
+			})
+			if m.ID > low {
+				low = m.ID
+			}
+		}
+		count = int32(len(batch))
+		glog.Infof("ReadLog(%v, (%v, %v], %v) count: %v", logID, low, high, batchSize, count)
+	}
+	return nil
+}
+
+// mapLogItem takes an individual entry and emits KV<index, mutation>
+// TODO(gbelvin): Make this mapping function part of the mutation interface.
+func mapLogItem(e *ktpb.EntryUpdate, emit func(index []byte, update *ktpb.EntryUpdate)) {
+	var entry ktpb.Entry
+	if err := proto.Unmarshal(e.GetMutation().GetEntry(), &entry); err != nil {
+		glog.Warningf("proto.Unmarshal failed: %v", err)
+	}
+	emit(entry.Index, e)
+}
+
+type mergeIndexFn struct{}
+
+func (*mergeIndexFn) CreateAccumulator() [][]byte                 { return [][]byte{} }
+func (*mergeIndexFn) AddInput(list [][]byte, val []byte) [][]byte { return append(list, val) }
+func (*mergeIndexFn) ExtractOutput(list [][]byte) [][]byte        { return list }
+func (*mergeIndexFn) MergeAccumulators(list [][][]byte) [][]byte {
+	ret := [][]byte{}
+	for _, l := range list {
+		ret = append(ret, l...)
+	}
+	return ret
+}
+
+// ReadMap queries the Trillian map for a list of leaves and emits KV<index, MapLeaf>
+func (s *Server) ReadMap(ctx context.Context, indexes [][]byte, directoryID string, rev int64,
+	emit func(index []byte, leaf *tpb.MapLeaf)) error {
+	// Fetch verification objects for directoryID.
+	config, err := s.ktServer.GetDirectory(ctx, &ktpb.GetDirectoryRequest{DirectoryId: directoryID})
+	if err != nil {
+		return err
+	}
+	mapClient, err := tclient.NewMapClientFromTree(s.tmap, config.Map)
+	if err != nil {
+		return err
+	}
+	leaves, err := mapClient.GetAndVerifyMapLeavesByRevision(ctx, rev, indexes)
+	if err != nil {
+		return err
+	}
+	for _, l := range leaves {
+		emit(l.Index, l)
+	}
+	return nil
+}
+
+// applyMutation processes all the mutations for a given index and emits the new map leaf.
+func applyMutation(index []byte,
+	getMapLeaf func(**tpb.MapLeaf) bool,
+	getMessage func(**ktpb.EntryUpdate) bool,
+	emit func(*tpb.MapLeaf)) error {
+	var oldMapLeaf *tpb.MapLeaf
+	if !getMapLeaf(&oldMapLeaf) {
+		return status.Errorf(codes.NotFound, "no map leaf found for index %x", index)
+	}
+	oldValue, err := entry.FromLeafValue(oldMapLeaf.LeafValue)
+	if err != nil {
+		glog.Warningf("entry.FromLeafValue(%v): %v", oldMapLeaf.LeafValue, err)
+		return err
+	}
+
+	var newEntry *ktpb.SignedEntry
+	var committed *ktpb.Committed
+	var msg *ktpb.EntryUpdate
+	for getMessage(&msg) {
+		// TODO(gbelvin): Modify mutation function to accept this function's signature.
+		newValue, err := entry.New().Mutate(oldValue, msg.Mutation)
+		if err != nil {
+			glog.Warningf("Mutate(): %v", err)
+			continue // Filter for valid mutations.
+		}
+		// The order of mutations needs to be associative we need to break ties without regard to the order of messages.
+		if bytes.Compare(newValue.Entry, newEntry.GetEntry()) > 0 {
```

Comment:
I wonder if this conflict should be resolved in favor of the "latest" mutation is some sense. Couple of valid reasons for having conflicts in the first place can be:

1. Client submits a mutation and fails / disconnects, then shortly after it submits a new mutation.
2. Client submits a mutation and says "oups, wrong one", then shortly after it submits a new one.

In both cases there is a chance for both mutations getting to the same revision.